### DISCUSSION TITLE
Buoyancy: Potential unit mismatch?

### DISCUSSION THREAD

**Discussion by RouvenZ** [lbl-srg/modelica-buildings#3457] [Q&A]
Dear all,

while delving into storage models I stumbled upon line [33](https://github.com/lbl-srg/modelica-buildings/blob/5c00f9da2dcaeec3e6facd8235544359f197dbb0/Buildings/Fluid/Storage/BaseClasses/Buoyancy.mo#L33C5-L33C5) in the [Buoyancy](https://github.com/lbl-srg/modelica-buildings/blob/5c00f9da2dcaeec3e6facd8235544359f197dbb0/Buildings/Fluid/Storage/BaseClasses/Buoyancy.mo)-Model. 
Here, `Q_flow[i]` is defined as `Q_flow[i] = k*noEvent(smooth(1, if dT[i]>0 then dT[i]^2 else 0));`. 
However, $\left [Q_{flow}[i] \right] = W$, $[k] = \frac{W}{K}$ and $[dT] = K$, so that the units seem to not check out here as $W \neq W\cdot K$. Am I overlooking something here, shouldn't there be a `sqrt` around the whole `noEvent(...)`? If there is indeed something missing, why does this not lead to an error while compiling?

As a side note: Could `k` be declared as a `Modelica.Units.SI.ThermalConductance` for readability purpose?

I am looking forward to your reply - have a nice day!

Cheers
Rouven

**Comment by mwetter** [comment]
The square term was added here:

> September 16, 2011 by Michael Wetter:
Changed the implementation from Q_flow[i] = k*max(heatPort[i+1].T-heatPort[i].T, 0); to Q_flow[i] = k*smooth(1, if dT[i]>0 then dT[i]^2 else 0);. The previous implementation was not differentiable. In modeling a solar system, this change reduced the computing time by a factor of 20 during the time when the pumps were almost switched off and colder temperature was fed from the collector to the tank.

The purpose of this model is to suppress a temperature inversion in the tank, and it does so with a differentiable term. This should not be considered as heat conduction, because the effect if there is a temperature inversion would be much more complicated. This model just brings this (physically unstable) configuration back to the point where there is no temperature inversion.

**Comment by RouvenZ** [comment]
Thank you for your reply, @mwetter! I had seen that in the documentation and understood that it was introduced to make the equation differentiable (I made the mistake of forgetting that `sqrt` would lead to it again not be differentiable at `dT == 0`). 
However, would this not be able to be solved with `smoothMax` instead (taking into account its overshooting behaviour from [IBPSA-issue 1726](https://github.com/ibpsa/modelica-ibpsa/issues/1726))? As I would understand the following part

> For simplicity, this model does not compute a buoyancy induced mass flow rate,
but rather a heat flow that has the same magnitude as the enthalpy flow
associated with the buoyancy induced mass flow rate.

I would assume the `Q_flow[i]` to be of the same magnitude as the (emulated) enthalpy flow if the mass flow were modelled. With the square-implementation, I would think that for `dT <1` one underestimates that enthalpy flow, for `dT > 1` one overestimates it. So I would expect the simulation to have an intermediate step where segments still end up with temperature inversion so that then the whole equation system including [Buoyancy](https://github.com/lbl-srg/modelica-buildings/blob/5c00f9da2dcaeec3e6facd8235544359f197dbb0/Buildings/Fluid/Storage/BaseClasses/Buoyancy.mo) is called again until there is no temperature inversion anymore, is that correct? If so, I assume this to at least happen less if one implements it with `smoothMax`, what do you think?

Regardless, I am confused that the unit checking does not throw a unit mismatch error - are `noEvent` somehow exempt from this, do you know?
